Compile wasmtime with effect-handlers/wasm-tools:func-ref-2

dhil opened the pull request:
This patch makes `wasmtime` compile with branch `func-ref-2` in our `wasm-tools` fork. There are many todos left to be completed.

The test suite compiles and passes.

dhil committed:
Make wasmtime-types type check

dhil committed:
Make wasmtime-environ type check.

dhil committed:
Make wasmtime-runtime type check

dhil committed:
Make cranelift-wasm type check

dhil committed:
Make wasmtime-cranelift type check

dhil committed:
Make wasmtime type check

dhil committed:
Make wasmtime-wast type check

fitzgen committed:
Use a `SmallVec` for `ABIArg`s (#4584)

Instead of a regular `Vec`.

These vectors are usually very small, for example here is the histogram of sizes
when running Sightglass's `pulldown-cmark` benchmark:

```
;; Number of samples = 10332
;; Min = 0
;; Max = 11
;;
;; Mean = 2.496128532713901
;; Standard deviation = 2.2859559855427243
;; Variance = 5.225594767838607
;;
;; Each ∎ is a count of 62
;;
 0 ..  1 [ 3134 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 1 ..  2 [ 2032 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 2 ..  3 [  159 ]: ∎∎
 3 ..  4 [  838 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎
 4 ..  5 [  970 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 5 ..  6 [ 2566 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 6 ..  7 [  303 ]: ∎∎∎∎
 7 ..  8 [  272 ]: ∎∎∎∎
 8 ..  9 [   40 ]:
 9 .. 10 [   18 ]:
```

By using a `SmallVec` with capacity of 6 we avoid the vast majority of heap
allocations and get some nice benchmark wins of up to ~1.11x faster compilation.

<h3>Sightglass Benchmark Results</h3>

```
compilation :: nanoseconds :: benchmarks/spidermonkey/benchmark.wasm

  Δ = 340361395.90 ± 63384608.15 (confidence = 99%)

  main.so is 0.88x to 0.92x faster than smallvec.so!
  smallvec.so is 1.09x to 1.13x faster than main.so!

  [3101467423 3425524333.41 4060621653] main.so
  [2820915877 3085162937.51 3375167352] smallvec.so

compilation :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  Δ = 988446098.59 ± 184075718.89 (confidence = 99%)

  main.so is 0.88x to 0.92x faster than smallvec.so!
  smallvec.so is 1.09x to 1.13x faster than main.so!

  [9006994951 9948091070.66 11792481990] main.so
  [8192243090 8959644972.07 9801848982] smallvec.so

compilation :: nanoseconds :: benchmarks/bz2/benchmark.wasm

  Δ = 7854567.87 ± 2215491.16 (confidence = 99%)

  main.so is 0.89x to 0.94x faster than smallvec.so!
  smallvec.so is 1.07x to 1.12x faster than main.so!

  [80354527 93864666.76 119789198] main.so
  [77554917 86010098.89 94726994] smallvec.so

compilation :: cycles :: benchmarks/bz2/benchmark.wasm

  Δ = 22810509.85 ± 6434024.63 (confidence = 99%)

  main.so is 0.89x to 0.94x faster than smallvec.so!
  smallvec.so is 1.07x to 1.12x faster than main.so!

  [233358190 272593088.57 347880715] main.so
  [225227821 249782578.72 275097380] smallvec.so

compilation :: nanoseconds :: benchmarks/pulldown-cmark/benchmark.wasm

  Δ = 10849521.41 ± 4324757.85 (confidence = 99%)

  main.so is 0.90x to 0.96x faster than smallvec.so!
  smallvec.so is 1.04x to 1.10x faster than main.so!

  [133875427 156859544.47 222455440] main.so
  [126073854 146010023.06 181611647] smallvec.so

compilation :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  Δ = 31508176.97 ± 12559561.91 (confidence = 99%)

  main.so is 0.90x to 0.96x faster than smallvec.so!
  smallvec.so is 1.04x to 1.10x faster than main.so!

  [388788638 455536988.31 646034523] main.so
  [366132033 424028811.34 527419755] smallvec.so
```

dhil committed:
Make testsuite compile

dhil committed:
Fix typo

dhil committed:
Fix test

fitzgen committed:
Use a `SmallVec` for `ABIArgSlot`s (#4586)

These are always length 1 for Wasm benchmarks.

<h3>Sightglass Benchmark Results</h3>

```
compilation :: nanoseconds :: benchmarks/spidermonkey/benchmark.wasm

  Δ = 328624015.86 ± 40274677.93 (confidence = 99%)

  main.so is 0.88x to 0.91x faster than slots-smallvec.so!
  slots-smallvec.so is 1.10x to 1.13x faster than main.so!

  [3070752447 3203778792.55 3446269274] main.so
  [2503544039 2875154776.69 3197966713] slots-smallvec.so

compilation :: nanoseconds :: benchmarks/pulldown-cmark/benchmark.wasm

  Δ = 9685705.06 ± 3221286.87 (confidence = 99%)

  main.so is 0.91x to 0.96x faster than slots-smallvec.so!
  slots-smallvec.so is 1.05x to 1.09x faster than main.so!

  [129356493 145594942.79 165038803] main.so
  [118555011 135909237.73 188780619] slots-smallvec.so

compilation :: nanoseconds :: benchmarks/bz2/benchmark.wasm

  No difference in performance.

  [79080493 86757564.46 112649639] main.so
  [78083384 85934125.69 94992743] slots-smallvec.so
```

alexcrichton committed:
Reduce wasm invocations in the `stacks` fuzzer (#4595)

On oss-fuzz a test case has been found that executes 30k iterations of a
wasm trap which with a 60s timeout leaves 2ms for each invocation which
under fuzzing instrumentation is a bit of a stretch with a ~20x
slowdown. This commit places a limit on the number of inputs to the
fuzzer at 200 to keep it reasonably sized.

alexcrichton committed:
Improve some support in `factc`: (#4580)

* Support CLI parameters for string encoding
* Fix `--skip-validate`
* Fix printing binary to stdout

alexcrichton committed:
Some minor cleanups/refactorings in components (#4582)

This is a collection of some minor renamings, refactorings, sharing of
code, etc. This was all discovered during my addition of string support
to adapter functions and I figured it'd be best to frontload this and
land it ahead of the full patch since it's getting complex.

alexcrichton committed:
Improve readability of memory64 compat in `fact` (#4581)

This commit aims to improve the readability of supporting the memory64
proposal in the `fact` adapter trampoline compiler. Previously there
were a few sprinkled blocks that used `if` to generate different
instructions inline, but as I've worked on support for strings this has
become pretty unwieldy as strings do far more memory manipulation than
other type conversions. A pattern that's easier to read is to have
small instruction helpers that take the pointer width as an argument and
internally dispatch to the correct instruction. This keeps the main
translation code branch-free and a bit easier to follow. Additionally
for more complicated branching logic it allows for deduplicating the
main translation path by having lots of little branches instead of one
large branch with everything duplicated on both halves.

afonso360 committed:
cranelift: Implement scalar FMA on x86 (#4460)

x86 does not have dedicated instructions for scalar FMA, lower
to a libcall which seems to be what llvm does.

akirilov-arm committed:
Initial back-edge CFI implementation (#3606)

Give the user the option to sign and to authenticate function
return addresses with the operations introduced by the Pointer
Authentication extension to the Arm instruction set architecture.

Copyright (c) 2021, Arm Limited.

uweigand committed:
[s390x, abi_impl] Support struct args using explicit pointers (#4585)

This adds support for StructArgument on s390x.  The ABI for this
platform requires that the address of the buffer holding the copy
of the struct argument is passed from caller to callee as hidden
pointer, using a register or overflow stack slot.

To implement this, I've added an optional "pointer" filed to
ABIArg::StructArg, and code to handle the pointer both in common
abi_impl code and the s390x back-end.

One notable change necessary to make this work involved the
"copy_to_arg_order" mechanism.  Currently, for struct args
we only need to copy the data (and that need to happen before
setting up any other args), while for non-struct args we only
need to set up the appropriate registers or stack slots.
This order is ensured by sorting the arguments appropriately
into a "copy_to_arg_order" list.

However, for struct args with explicit pointers we need to *both*
copy the data (again, before everything else), *and* set up a
register or stack slot.  Since we now need to touch the argument
twice, we cannot solve the ordering problem by a simple sort.
Instead, the abi_impl common code now provided *two* callbacks,
emit_copy_regs_to_buffer and emit_copy_regs_to_arg, and expects
the back end to first call copy..to_buffer for all args, and
then call copy.._to_arg for all args.  This required updates
to all back ends.

In the s390x back end, in addition to the new ABI code, I'm now
adding code to actually copy the struct data, using the MVC
instruction (for small buffers) or a memcpy libcall (for larger
buffers).  This also requires a bit of new infrastructure:
- MVC is the first memory-to-memory instruction we use, which
  needed a bit of memory argument tweaking
- We also need to set up the infrastructure to emit libcalls.

(This implements the first half of issue #4565.)

elliottt committed:
x64: Begin migrating branch instructions to ISLE (#4587)

https://github.com/bytecodealliance/wasmtime/pull/4587

CosineP commented on cranelift/wasm/src/translation_utils.rs:
```diff
@@ -123,12 +144,18 @@ pub fn block_with_params<PE: TargetEnvironment + ?Sized>(
+            wasmparser::ValType::Ref(rt) => {
+                match rt.heap_type {
+                    wasmparser::HeapType::Func | wasmparser::HeapType::Extern => {
+                        builder.append_block_param(block, environ.reference_type(rt.heap_type.try_into()?));
```
surely the reference type constructor will need to know nullability as well?

CosineP commented on crates/cranelift/src/func_environ.rs:
```diff
@@ -1317,11 +1317,12 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
-            WasmType::ExternRef,
-            "We only use GlobalVariable::Custom for externref"
-        );
+        // TODO(dhil) fixme: need to check that wasm_ty is indeed a Ref.
```
I don't know anything about GlobalVariable::Custom, but if the message is true we can probably leave this as is but with EXTERN_REF

CosineP commented on crates/cranelift/src/lib.rs:
```diff
@@ -206,10 +207,10 @@ fn func_signature(
+        cranelift_wasm::WasmHeapType::Extern => match pointer_type {
             ir::types::I32 => ir::types::R32,
             ir::types::I64 => ir::types::R64,
             _ => panic!("unsupported pointer type"),
```
presumably need a case for Index

CosineP commented on crates/environ/src/module.rs:
```diff
@@ -428,7 +435,7 @@ impl ModuleTranslation<'_> {
             // If this is not a funcref table, then we can't support a
             // pre-computed table of function indices.
-            if self.module.table_plans[segment.table_index].table.wasm_ty != WasmType::FuncRef {
+            if !is_func_ref(self.module.table_plans[segment.table_index].table.wasm_ty) {
```
You can use == here just fine, just compare to WasmHeapType::Func

CosineP commented on crates/runtime/src/table.rs:
```diff
@@ -163,11 +163,11 @@ pub enum Table {
+    match rt.heap_type {
+        WasmHeapType::Func => Ok(TableElementType::Func),
+        WasmHeapType::Extern => Ok(TableElementType::Extern),
+        ht => bail!("invalid table element type {:?}", ht),
```
Need (a todo for) Index

CosineP commented on crates/types/src/lib.rs:
```diff
@@ -87,10 +161,10 @@ pub struct WasmFuncType {
     #[inline]
     pub fn new(params: Box<[WasmType]>, returns: Box<[WasmType]>) -> Self {
-        let externref_params_count = params.iter().filter(|p| **p == WasmType::ExternRef).count();
+        let externref_params_count = params.iter().filter(|p| match **p { WasmType::Ref(WasmRefType { heap_type: WasmHeapType::Extern, .. }) => true, _ => false }).count();
```
`p.heap_type == WasmHeapType::Extern` - even more succinct than the old version

CosineP commented on crates/types/src/lib.rs:
```diff
@@ -87,10 +161,10 @@ pub struct WasmFuncType {
         let externref_returns_count = returns
             .iter()
-            .filter(|r| **r == WasmType::ExternRef)
+            .filter(|r| match **r { WasmType::Ref(WasmRefType { heap_type: WasmHeapType::Extern, .. }) => true, _ => false })
```
Speaking of which do non-nullable extern refs exist?

dhil replied:
I don't think so, as `externref` and `funcref` are interpreted as nullable refs c.f. https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#Reference-Types.

CosineP commented on crates/wasmtime/src/externals.rs:
```diff
@@ -459,7 +465,7 @@ impl Table {
     fn _new(store: &mut StoreOpaque, ty: TableType, init: Val) -> Result<Table> {
         let wasmtime_export = generate_table_export(store, &ty)?;
-        let init = init.into_table_element(store, ty.element())?;
+        let init = init.into_table_element(store, ValType::Ref(ty.element()))?;
```
maybe into_table_element should accept a RefType?

CosineP commented on crates/wasmtime/src/func/typed.rs:
```diff
@@ -321,7 +321,7 @@ unsafe impl WasmTy for Option<ExternRef> {
 
```
Where does ExternRef come from here?

dhil replied:
It is defined in https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasmtime/src/ref.rs#L10

CosineP commented on crates/wasmtime/src/types.rs:
```diff
@@ -47,8 +47,10 @@ impl fmt::Display for ValType {
             ValType::V128 => write!(f, "v128"),
-            ValType::ExternRef => write!(f, "externref"),
-            ValType::FuncRef => write!(f, "funcref"),
+            ValType::Ref(rt) => write!(f, "ref {}", rt),
```
"ref " should presumably be handled by Ref printing (and parens)

CosineP commented on crates/wasmtime/src/types.rs:
```diff
@@ -90,14 +92,88 @@ impl ValType {
+// Reference types
+/// A list of all possible value types in WebAssembly.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct RefType {
```
Just curious, why does this get defined twice?

CosineP reviewed (approved):
This is great, and looks like the tests pass too?   Cool.  Alright looks like we have our work cut out for us :)

jameysharp committed:
Upgrade regalloc2 -> 0.3.2 (#4603)

Includes a modest improvement in memory usage and performance by
removing analysis that was only used during fuzzing.

fitzgen committed:
Save exit Wasm FP and PC in component-to-host trampolines  (#4601)

* Wasmtime: Add a pointer to `VMRuntimeLimits` in component contexts

* Save exit Wasm FP and PC in component-to-host trampolines

Fixes #4535

* Add comment about why we deref the trampoline's FP

* Update some tests to use new `vmruntime_limits_*` methods

dheaton-arm committed:
Port `Shuffle` to ISLE (AArch64) (#4596)

* Port `Shuffle` to ISLE (AArch64)

Ported the existing implementation of `Shuffle` for AArch64 to ISLE.

Copyright (c) 2022 Arm Limited

* Cleanup by shadowing `rn`, `rn2`, and `_`

Copyright (c) 2022 Arm Limited

elliottt committed:
x64: Migrate brff and I128 branching instructions to ISLE (#4599)

https://github.com/bytecodealliance/wasmtime/pull/4599

teymour-aldridge committed:
Add `try_use_var` method to `cranelift-frontend`. (#4588)

* Add `try_use_var` method to `cranelift-frontend`.
- Unlike `use_var`, this method does not panic if the variable has not been defined
before use

* Add `try_declare_var` and `try_def_var`.
- Also implement Error for error enums.

* Use `write!` macro.

* Add `write!` use I missed.

dicej committed:
implement fuzzing for component types (#4537)

This addresses #4307.

For the static API we generate 100 arbitrary test cases at build time, each of
which includes 0-5 parameter types, a result type, and a WAT fragment containing
an imported function and an exported function.  The exported function calls the
imported function, which is implemented by the host.  At runtime, the fuzz test
selects a test case at random and feeds it zero or more sets of arbitrary
parameters and results, checking that values which flow host-to-guest and
guest-to-host make the transition unchanged.

The fuzz test for the dynamic API follows a similar pattern, the only difference
being that test cases are generated at runtime.

Signed-off-by: Joel Dice <joel.dice@fermyon.com>

elliottt committed:
x64: Finish migrating brz and brnz to ISLE (#4614)

https://github.com/bytecodealliance/wasmtime/pull/4614

uweigand committed:
[s390x, abi_impl] Add i128 support (#4598)

This adds full i128 support to the s390x target, including new filetests
and enabling the existing i128 runtest on s390x.

The ABI requires that i128 is passed and returned via implicit pointer,
but the front end still generates direct i128 types in call.  This means
we have to implement ABI support to implicitly convert i128 types to
pointers when passing arguments.

To do so, we add a new variant ABIArg::ImplicitArg.  This acts like
StructArg, except that the value type is the actual target type,
not a pointer type.  The required conversions have to be inserted
in the prologue and at function call sites.

Note that when dereferencing the implicit pointer in the prologue,
we may require a temp register: the pointer may be passed on the
stack so it needs to be loaded first, but the value register may
be in the wrong class for pointer values.  In this case, we use
the "stack limit" register, which should be available at this
point in the prologue.

For return values, we use a mechanism similar to the one used for
supporting multiple return values in the Wasmtime ABI.  The only
difference is that the hidden pointer to the return buffer must
be the *first*, not last, argument in this case.

(This implements the second half of issue #4565.)

alexcrichton committed:
Add a dataflow-based representation of components (#4597)

* Add a dataflow-based representation of components

This commit updates the inlining phase of compiling a component to
creating a dataflow-based representation of a component instead of
creating a final `Component` with a linear list of initializers. This
dataflow graph is then linearized in a final step to create the actual
final `Component`.

The motivation for this commit stems primarily from my work implementing
strings in fused adapters. In doing this my plan is to defer most
low-level transcoding to the host itself rather than implementing that
in the core wasm adapter modules. This means that small
cranelift-generated trampolines will be used for adapter modules to call
which then call "transcoding libcalls". The cranelift-generated
trampolines will get raw pointers into linear memory and pass those to
the libcall which core wasm doesn't have access to when passing
arguments to an import.

Implementing this with the previous representation of a `Component` was
becoming too tricky to bear. The initialization of a transcoder needed
to happen at just the right time: before the adapter module which needed
it was instantiated but after the linear memories referenced had been
extracted into the `VMComponentContext`. The difficulty here is further
compounded by the current adapter module injection pass already being
quite complicated. Adapter modules are already renumbering the index
space of runtime instances and shuffling items around in the
`GlobalInitializer` list. Perhaps the worst part of this was that
memories could already be referenced by host function imports or exports
to the host, and if adapters referenced the same memory it shouldn't be
referenced twice in the component. This meant that `ExtractMemory`
initializers ideally needed to be shuffled around in the initializer
list to happen as early as possible instead of wherever they happened to
show up during translation.

Overall I did my best to implement the transcoders but everything always
came up short. I have decided to throw my hands up in the air and try a
completely different approach to this, namely the dataflow-based
representation in this commit. This makes it much easier to edit the
component after initial translation for injection of adapters, injection
of transcoders, adding dependencies on possibly-already-existing items,
etc. The adapter module partitioning pass in this commit was greatly
simplified to something which I believe is functionally equivalent but
is probably an order of magnitude easier to understand.

The biggest downside of this representation I believe is having a
duplicate representation of a component. The `component::info` was
largely duplicated into the `component::dfg` module in this commit.
Personally though I think this is a more appropriate tradeoff than
before because it's very easy to reason about "convert representation A
to B" code whereas it was very difficult to reason about shuffling
around `GlobalInitializer` items in optimal fashions. This may also have
a cost at compile-time in terms of shuffling data around, but my hope is
that we have lots of other low-hanging fruit to optimize if it ever
comes to that which allows keeping this easier-to-understand
representation.

Finally, to reiterate, the final representation of components is not
changed by this PR. To the runtime internals everything is still the
same.

* Fix compile of factc

elliottt committed:
x64: Migrate br_table to ISLE (#4615)

https://github.com/bytecodealliance/wasmtime/pull/4615

Commit:
components: ignore export aliases to types in translation. (#4604)

* components: ignore export aliases to types in translation.

Currently, translation is ignoring type exports from components during
translation by skipping over them before adding them to the exports map.

If a component instantiates an inner component and aliases a type export of
that instance, it will cause wasmtime to panic with a failure to find the
export in the exports map.

The fix is to add a representation for exported types to the map that is simply
ignored when encountered. This also makes it easier to track places where we
would have to support type exports in translation in the future.

* Keep type information for type exports.

This commit keeps the type information for type exports so that types can be
properly aliased from an instance export and thereby adjusting the type index
space accordingly.

* Add a simple test case for type exports for the component model.

uweigand committed:
s390x: Implement bitrev (#4617)

Since we do not have an instruction for this, this is a simple
open-coded implementation.

Needed by the cg_clif frontend.

wasmtime-publish committed:
Bump Wasmtime to 0.41.0 (#4620)

Co-authored-by: Wasmtime Publish <wasmtime-publish@users.noreply.github.com>

alexcrichton committed:
Fix an issue in adapter module partitioning (#4622)

When an adapter module depends on a particular core wasm instance this
means that it actually depends on not only that instance but all prior
core wasm instances as well. This is because core wasm instances must be
instantiated in the specified order within a component and that cannot
change depending on the dataflow between adapters. This commit fixes a
possible panic from linearizing the component dfg where an adapter
module tried to depend on an instance that hadn't been instantiated yet
because the ordering dependency between core wasm instances hadn't been
modeled.

dhil committed:
Fix TODOs for debug_assert_eq!

dhil committed:
Address Luna's comments

dhil committed:
Merge with upstream
